crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs

```diff
@@ -319,62 +286,120 @@ impl Inner {
         Ok(scheduler)
     }
 
-    fn step<C>(&mut self, circuit: &C) -> Result<(), Error>
+    /// Spawn a tokio task to evaluate `node_id`.
+    fn spawn_task<C>(&mut self, circuit: &C, node_id: NodeId)
     where
         C: Circuit,
     {
-        circuit.log_scheduler_event(&SchedulerEvent::step_start(circuit.global_id().deref()));
+        let task = &mut self.tasks[node_id.id()];
+        debug_assert!(task.unsatisfied_dependencies == 0);
+        debug_assert!(task.is_ready);
+        debug_assert!(!task.scheduled);
+
+        task.scheduled = true;
+
+        if self.handles.is_empty() && self.waiting {
+            self.waiting = false;
+            circuit.log_scheduler_event(&SchedulerEvent::wait_end(circuit.global_id()));
+        }
+
+        let circuit = circuit.clone();
+
+        self.handles.spawn_local(async move {
+            let result = circuit.eval_node(node_id).await;
+            (node_id, result)
+        });
+    }
+
+    async fn abort(&mut self) {
+        // wait for all started tasks to abort
```

Comment:
We could cancel the tasks instead of just waiting for them to finish, by adding:

```
        self.handles.abort_all();
```

Reply:
I wasn't sure what was the best choice here. It's OK with me not to cancel them.